course/api/lti_views.py

```diff
@@ -0,0 +1,127 @@
+    permission_classes = [OAuth2TokenPermission]
+
+    def check_token_scope(self, scopes: List[str]) -> bool:
+        if self.action == 'scores':
+            if 'https://purl.imsglobal.org/spec/lti-ags/scope/score' in scopes:
+                return True
+        else:
+            if 'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly' in scopes:
+                return True
+        return False
+
+    def list(self, request: Request, *args, **kwargs):
+        queryset = self.queryset.filter(course_module__course_instance__pk=kwargs['course_id'])
+        result = []
+        for exercise in queryset:
+            lineitem = LineItem(request, exercise, kwargs['course_id'])
+            serializer = LtiLineItemSerializer(lineitem)
+            result.append(serializer.data)
+
+        return Response(result)
+
+    def retrieve(self, request: Request, *args, **kwargs):
+        try:
+            exercise = LTI1p3Exercise.objects.get(pk=kwargs['exercise_id'])
+            lineitem = LineItem(request, exercise, kwargs['course_id'])
+            serializer = LtiLineItemSerializer(lineitem)
+        except LTI1p3Exercise.DoesNotExist as e:
+            logger.error("LTI 1.3/lineitem: Exercise (%d) not found", kwargs['exercise_id'])
+            raise Http404("LTI exercise not found") from e
+        return Response(serializer.data)
+
+    @action(methods=['post'], detail=True, url_path='scores')
+    def scores(self, request: Request, *args, **kwargs):
+        jsondata = json.loads(request.body)
+        serializer = LtiScoresSerializer(data=jsondata)
+        serializer.is_valid()
+        data = serializer.validated_data
+
+        try:
+            exercise = LTI1p3Exercise.objects.get(pk=kwargs['exercise_id'])
+        except LTI1p3Exercise.DoesNotExist as e:
+            logger.error("LTI 1.3/scores: Exercise (%d) not found", kwargs['exercise_id'])
+            raise Http404("LTI exercise not found") from e
+
+        try:
+            user = User.objects.get(pk=int(data.get('userId')))
+        except UserProfile.DoesNotExist:
+            logger.error("LTI 1.3/scores: User (%s) does not exist", data.get('userId'))
+            return Response({'UserProfile does not exist'}, status=400)
+
+        logger.info("Received LTI 1.3 'scores' request for exercise %s, user %s", str(exercise), str(user))
+        sub = Submission.objects.filter(exercise=exercise, submitters=user.userprofile)
+        if sub.exists():
+            # When points come from LTI, we assume there is only one submission per exercise per user,
+            # that can be updated if LTI delivers larger points than previously.
+            # If points are not larger than previously, we will just ignore the message.
+            # Some LTI tools seem to send scores updates quite frequently.
+            sub = sub.first()
+            adjusted = (1.0 * exercise.max_points * data.get('scoreGiven') / data.get('scoreMaximum'))
```

Comment:
Have `scoreGiven` and `scoreMaximum` been validated at this point so that they are floats and not any nonsensical types?